projects/client/RabbitMQ.Client/src/client/impl/SocketFrameHandler.cs

```diff
@@ -1,321 +1,304 @@
+//  Copyright (c) 2007-2016 Pivotal Software, Inc.  All rights reserved.
+//---------------------------------------------------------------------------
+
+#if !NETFX_CORE
+using RabbitMQ.Client.Exceptions;
+using RabbitMQ.Util;
+using System;
+using System.IO;
+using System.Net;
+using System.Collections.Generic;
+using System.Net.Sockets;
+using System.Text;
+using System.Threading.Tasks;
+using System.Threading;
+
+namespace RabbitMQ.Client.Impl
+{
+    static class TaskExtensions
+    {
+        public static Task CompletedTask = Task.FromResult(0);
+
+        public static async Task TimeoutAfter(this Task task, int millisecondsTimeout)
+        {
+            if (task == await Task.WhenAny(task, Task.Delay(millisecondsTimeout)).ConfigureAwait(false))
+                await task;
+            else
+                throw new TimeoutException();
+        }
+    }
+
+    public class SocketFrameHandler : IFrameHandler
+    {
+        // Timeout in seconds to wait for a clean socket close.
+        private const int SOCKET_CLOSING_TIMEOUT = 1;
+        // Socket poll timeout in ms. If the socket does not
+        // become writeable in this amount of time, we throw
+        // an exception.
+        private int m_writeableStateTimeout = 30000;
+        private readonly NetworkBinaryReader m_reader;
+        private readonly ITcpClient m_socket;
+        private readonly NetworkBinaryWriter m_writer;
+        private readonly object _semaphore = new object();
+        private bool _closed;
+        public SocketFrameHandler(AmqpTcpEndpoint endpoint,
+            Func<AddressFamily, ITcpClient> socketFactory,
+            int connectionTimeout, int readTimeout, int writeTimeout)
+        {
+            Endpoint = endpoint;
+
+            if (ShouldTryIPv6(endpoint))
+            {
+                try {
+                    m_socket = ConnectUsingIPv6(endpoint, socketFactory, connectionTimeout);
+                } catch (ConnectFailureException)
+                {
+                    m_socket = null;
+                }
+            }
+
+            if (m_socket == null && endpoint.AddressFamily != AddressFamily.InterNetworkV6)
+            {
+                m_socket = ConnectUsingIPv4(endpoint, socketFactory, connectionTimeout);
+            }
+
+            Stream netstream = m_socket.GetStream();
+            netstream.ReadTimeout  = readTimeout;
+            netstream.WriteTimeout = writeTimeout;
+
+            if (endpoint.Ssl.Enabled)
+            {
+                try
+                {
+                    netstream = SslHelper.TcpUpgrade(netstream, endpoint.Ssl);
+                }
+                catch (Exception)
+                {
+                    Close();
+                    throw;
+                }
+            }
+            m_reader = new NetworkBinaryReader(new BufferedStream(netstream, m_socket.Client.ReceiveBufferSize));
+            m_writer = new NetworkBinaryWriter(netstream);

```

Comment:
Removed use of bufferedstream on the write side.
Information was being written to the stream and flushed immediately.
The change to chunk information onto the stream means we don't have to buffer.
Since we don't have to buffer, we also don't need to lock the flush.
The write doesn't need to be locked when done as a chunk.